Subject: Fix extraction of HID items >= 32 bit in size
Date: Sun Mar 13 22:11:38 CET 2005
From: Vojtech Pavlik <vojtech@suse.cz>
===================================================================

ChangeSet@1.2073, 2005-03-13 20:19:37+01:00, akropel1@rochester.rr.com
  input: HID items of width 32 (bits) or greater are incorrectly extracted due to
  a masking bug in hid-core.c:extract(). This patch fixes it up by forcing
  the mask to be 64 bits wide.
  
  Signed-off-by: Adam Kropelin <akropel1@rochester.rr.com>
  Signed-off-by: Vojtech Pavlik <vojtech@suse.cz>

 hid-core.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -Nru a/drivers/usb/input/hid-core.c b/drivers/usb/input/hid-core.c
--- a/drivers/usb/input/hid-core.c	2005-03-13 22:11:43 +01:00
+++ b/drivers/usb/input/hid-core.c	2005-03-13 22:11:43 +01:00
@@ -766,7 +766,7 @@
 static __inline__ __u32 extract(__u8 *report, unsigned offset, unsigned n)
 {
 	report += (offset >> 5) << 2; offset &= 31;
-	return (le64_to_cpu(get_unaligned((__le64*)report)) >> offset) & ((1 << n) - 1);
+	return (le64_to_cpu(get_unaligned((__le64*)report)) >> offset) & ((1ULL << n) - 1);
 }
 
 static __inline__ void implement(__u8 *report, unsigned offset, unsigned n, __u32 value)
